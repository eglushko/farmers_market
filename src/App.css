#root {
    padding-top: 5rem;
}

.card {
    border-radius: 10px;
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.25);
}

button {
    background-color: #83072c; 
    color: white;
    cursor: pointer;
    border: solid 1px #2c0d00;
    border-radius: 8px; 
    font-family: sans-serif;
    font-size: 1rem;
    font-weight: bold;
    white-space: nowrap;
    overflow: hidden;
}

button:hover,
button:active {
    color: #83072c;
    background-color: white;
    border-color: #83072c; 
}
