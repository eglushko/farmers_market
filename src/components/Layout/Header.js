import { Fragment, useContext } from 'react';

import cartIcon from '../../assets/shopping-cart-icon-48.png';
import CartContext from '../../store/CartContext';
import classes from './Header.module.css';


const Header = (props) => {
    const cartCtx = useContext(CartContext);
    const numOfCartItems = cartCtx.cartItems.reduce((curNum, item) => {
        return curNum + item.quantity;
    }, 0);
    return (
        <Fragment>
            <header className={classes.header}>
                <h1>🍎 <span className={classes.title}>Farmers Market</span></h1>
                <button className={classes.button} onClick={props.onOpenCart}>
                    <span className={classes.icon}><img src={cartIcon} alt="Cart" /></span>
                    <span>Your Cart</span>
                    <span className={classes.badge}>{numOfCartItems}</span>
                </button>
            </header>
        </Fragment>
    );

};

export default Header;
