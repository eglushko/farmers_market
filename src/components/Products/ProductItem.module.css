.product-item {
    display: flex;
    height: 100%;
    flex-direction: column;
    width: 20%;
    min-width: 260px;
    max-width: 320px;
    margin: 1rem;
    padding: 0.5rem 0.5rem 1rem 0.5rem;
    border: 1px solid #ccc;
    box-shadow: 0 1px 2px 0 #d8d8d8;
    border-radius: 4px;
    background-color: #ededed;
}
  
.image-container {
    display: flex;
    width: 100%;
    flex: 1;
    border: solid 2px #ededed;
    cursor: pointer;
    border-radius: 4px;
}

.image-container:hover {
    border-color: #8a2b06;
}
  
.image-container img {
    width: 100%;
    border-radius: 4px;
}
  
.product-item h3 {
    cursor: pointer;
    margin: 0.5rem 0 0.25rem 0;
    flex: 1;
}

.product-item h3:hover  {
    text-decoration: underline;
}
  
.product-item-description {
    font-style: italic;
    min-height: 57px;
    flex: 1;
}
  
.product-item-price {
    margin-top: 0.25rem;
    font-weight: bold;
    font-size: 1.25rem;
    flex: 1;
}
  
.form {
    margin-top: 1rem;
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
}

.form input {
    margin-top: 0.4rem;
}

.form input label {
    margin-right: 0.5rem;
}
  
.form button {
    padding: 0.25rem 0.5rem;
    margin-left: 0.5rem;
    border-radius: 6px;
}

.availability {
    color: red;
    margin: 1rem 0 1rem 0;
    font-size: 1rem;
    font-weight: bold;
}

@media only screen and (max-width: 300px) {
    .product-item {
        min-width: 200px;
        margin: 0.4rem;
    }  
    
    .long {
        display: none;
    }
}