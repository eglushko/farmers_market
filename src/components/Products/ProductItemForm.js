import { useRef, useState } from 'react';
import Input from '../UI/Input';

import classes from './ProductItem.module.css';

const ProductItemForm = (props) => {
    const quantityInputRef = useRef();
    const [quantityIsValid, setQuantityIsValid] = useState(true);
    const formSubmitHandler = (e) => {
        e.preventDefault();
        const enteredQuantity = quantityInputRef.current.value;
        if (enteredQuantity.trim().length === 0 || enteredQuantity < 1) {
            setQuantityIsValid(false);
            return;
        }
        const enteredQuantityNumber = +enteredQuantity;
        props.onAddToCart(enteredQuantityNumber);
    }
    return (
        <form className={classes.form} onSubmit={formSubmitHandler}>
            <Input label="Quantity" ref={quantityInputRef} input={{
                id: 'quantity_' + props.id, 
                type: 'number',
                min: '1',
                max: '100',
                step: '1',
                defaultValue: '1'
                }} />
            {!quantityIsValid && <p>Quantity is invalid.</p>}
            <button>Add <span className={classes.long}>to Cart</span></button>
        </form>
    );
};

export default ProductItemForm;

