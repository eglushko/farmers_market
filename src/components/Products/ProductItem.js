import { useContext } from 'react';
import ProductItemForm from './ProductItemForm';
import CartContext from '../../store/CartContext';

import classes from './ProductItem.module.css';

const ProductItem = (props) => {
    const price = props.item.price.toFixed(2);
    const cartCtx = useContext(CartContext);
    const addToCartHandler = (quantity) => {
        cartCtx.addItem({
            id: props.item.id,
            image: props.item.image,
            name: props.item.name,
            quantity: quantity,
            price: props.item.price
          });
    };
    const openProductPageHandler = () => {
        props.onOpenProductPage(props.item);
    };
    return (
        <li className={classes['product-item']}>
            <div className={classes['image-container']} onClick={openProductPageHandler}>
                <img src={props.item.image} alt={props.item.name} />
            </div>
            <h3 onClick={openProductPageHandler}>{props.item.name}</h3>
            <div className={classes['product-item-price']}>${price}</div>
            {props.item.isAvailable && <ProductItemForm id={props.item.id} onAddToCart={addToCartHandler} />}
            {!props.item.isAvailable && <div className={classes.availability}>Currently unavailable</div>}
        </li>
    );
};

export default ProductItem;

