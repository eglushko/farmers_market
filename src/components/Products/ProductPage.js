import { useContext } from 'react';
import ProductItemForm from './ProductItemForm';
import CartContext from '../../store/CartContext';
import Modal from '../UI/Modal';
import classes from './ProductPage.module.css';

const ProductPage = (props) => {
    const item = props.item;
    const price = props.item.price.toFixed(2);
    const cartCtx = useContext(CartContext);
    const addToCartHandler = (quantity) => {
        const item = props.item;
        item.quantity = quantity;
        cartCtx.addItem(item);
    };
    return (
        <Modal item={item} onClose={props.onClose}>
            <div className={classes.product}>
                <div className={classes.main}>
                    <div className={classes.image}>
                        <img src={item.image} alt={item.name} />
                    </div>
                    <div className={classes.info}>
                        <div className={classes.name}>
                            <h2 title={item.name}>{item.name}</h2>
                        </div>
                        <div><h2>${price}</h2></div>
                        <div className={classes.rating}>
                            <div className={classes.text}>{item.rating} out of 5</div>
                            <div className={classes.stars} style={{'--rating': item.rating}} aria-label="Rating of this product is 2.3 out of 5."></div>
                        </div>
                        {!props.item.isAvailable && <div className={classes.availability}>Currently unavailable</div>}
                    </div>
                </div>
                <div className={classes.details}>
                    <div>{item.description}</div>
                </div>
                <div className={classes.actions}>
                    {props.item.isAvailable && <ProductItemForm id={props.item.id} onAddToCart={addToCartHandler} />}
                    <button className={classes.close} onClick={props.onClose}>Close</button>
                </div>
            </div>
        </Modal>
    );
};

export default ProductPage;