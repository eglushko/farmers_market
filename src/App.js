import React from 'react';
import { useState } from 'react';

import Cart from './components/Cart/Cart'
import Header from './components/Layout/Header';
import Products from './components/Products/Products';
import CartProvider from './store/CartProvider';

import './App.css';

const App = () => {
  const [cartIsOpen, setCartIsOpen] = useState(false);

  const openCartHandler = () => {
    setCartIsOpen(true);
  }

  const closeCartHandler = () => {
    setCartIsOpen(false);
  }

  return (
    <CartProvider>
      {cartIsOpen && <Cart onClose={closeCartHandler} />} 
      <Header onOpenCart={openCartHandler}></Header>
      <Products />
    </CartProvider> 
  );
};

export default App;
