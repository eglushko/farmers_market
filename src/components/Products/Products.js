import { useEffect, useState, useCallback, Fragment } from 'react';

import ProductsList from './ProductsList';
import ProductPage from './ProductPage';

import classes from './Products.module.css';

const Products = (props) => {
    const fetchProductsCall = './data/products.json'; 
    const [products, setProducts] = useState([]);
    const [isLoading, setIsLoading] = useState(false);
    const [error, setError] = useState(null);
    const [productPageState, setProductPageState] = useState({
        isOpen: false,
        activeItem: null
    });
    const openProductPageHandler = (item) => {
        setProductPageState({
            isOpen: true,
            activeItem: item
        });
    };

    const fetchProducts = useCallback(async () => {
        setIsLoading(true);
        setError(null);

        try {
              fetch(fetchProductsCall)    
                .then((res) => res.json())
                .then((data) => {
                    setIsLoading(false);
                    setProducts(data);
                });
        }
        catch (error) {
            setError(error.message);
        }
    }, []);

    useEffect(() => {
        let isApiSubscribed = true;
        fetchProducts();
        return () => {
            isApiSubscribed = false;
        }
    }, [fetchProducts]);

    const closeProductPageHandler = () => {
        setProductPageState({
            isOpen: false,
            activeItem: null    
        });
    };


    let content = <p className={classes.message}>No products found.</p>;
    if (products.length > 0) {
        content = <ProductsList items={products}  onOpenProductPage={openProductPageHandler} />;
    }
    
    if (error) {
        content = <p className={classes.message}>{error}</p>;
    }
    
    if (isLoading) {
        content = <div className={classes.message}>Loading...</div>;
    }
        
    return (
        <Fragment>
            {productPageState.isOpen && <ProductPage 
                item={productPageState.activeItem} 
                onClose={closeProductPageHandler} />}
            {content}
        </Fragment>
    );
};

export default Products;