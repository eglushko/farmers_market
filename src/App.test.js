import { render } from '@testing-library/react';
import App from './App';
import Header from './components/Layout/Header';
import Products from './components/Products/Products';

describe('App component', () => {
  test('Header component rendered properly', () => {
    const tree = render(
      <App>
      <Header />
      <Products />
      </App>
    ); 
    expect(tree).toMatchSnapshot();
  });
});
