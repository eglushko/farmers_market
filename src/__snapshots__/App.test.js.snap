// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component Header component rendered properly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="header"
      >
        <h1>
          üçé 
          <span
            class="title"
          >
            Sweet Apple Acres
          </span>
        </h1>
        <button
          class="button"
        >
          <span
            class="icon"
          >
            <img
              alt="Cart"
              src="shopping-cart-icon-48.png"
            />
          </span>
          <span>
            Your Cart
          </span>
          <span
            class="badge"
          >
            0
          </span>
        </button>
      </header>
      <div
        class="message"
      >
        Loading...
      </div>
    </div>
  </body>,
  "container": <div>
    <header
      class="header"
    >
      <h1>
        üçé 
        <span
          class="title"
        >
          Sweet Apple Acres
        </span>
      </h1>
      <button
        class="button"
      >
        <span
          class="icon"
        >
          <img
            alt="Cart"
            src="shopping-cart-icon-48.png"
          />
        </span>
        <span>
          Your Cart
        </span>
        <span
          class="badge"
        >
          0
        </span>
      </button>
    </header>
    <div
      class="message"
    >
      Loading...
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
